SECEdgar/filings/base.py

```diff
@@ -43,75 +44,98 @@ def dateb(self, val):
 
     @property
     def filing_type(self):
-        return self._validate_filing_type(self._filing_type)
+        return Filing._validate_filing_type(self._filing_type)
 
     @filing_type.setter
     def filing_type(self, ft):
         self._filing_type = self._validate_filing_type(ft)
 
     @property
     def cik(self):
-        return self._validate_cik(self._cik)
+        return Filing._validate_cik(self._cik)
 
-    @cik.setter
-    def cik(self, val):
-        self._cik = self._validate_cik(val)
-
-    def _validate_filing_type(self, filing_type):
+    @staticmethod
+    def _validate_filing_type(filing_type):
         """Validates that given filing type is valid.
 
         Args:
-            filing_type (str): Valid filing type (case-insensitive).
+            filing_type (filings.FilingType): Valid filing type enum.
 
         Raises:
             FilingTypeError: If filing type is not supported/valid.
 
         Returns:
-            filing_type (filing_types.FilingType): If filing type is valid, given filing
+            filing_type (filings.FilingType): If filing type is valid, given filing
                 type will be returned.
         """
         if not isinstance(filing_type, FilingType):
             raise FilingTypeError(FilingType)
 
         return filing_type
 
-    def _validate_cik(self, cik):
+    @staticmethod
+    def _validate_cik(cik):
         """Validates that given CIK *could* be valid.
 
         Args:
-            cik (Union[str, int]): Central index key (CIK) to validate.
+            cik (Union[CIK, str, int]): Central index key (CIK) to validate.
 
         Returns:
-            cik (Union[str, int]): Validated CIK.
+            cik (Union[str, list of str]): Validated CIK.
                 Note that the CIK is only validated in
-                that it *could* be valid. All CIKs
-                must be 10 digits, but not all 10 digit
+                that it *could* be valid. CIKs formatted as
+                10 digits, but not all 10 digit
                 numbers are valid CIKs.
 
         Raises:
-            ValueError: If given cik is not str or int
-            CIKError: If cik is not a 10 digit number.
+            ValueError: If given cik is not str, int, or CIK object.
+            CIKError: If cik is not a 10 digit number or valid CIK object
         """
-        if not isinstance(cik, (str, int)):
-            raise ValueError("CIK must be of type str or int.")
-        elif isinstance(cik, str):
-            if len(cik) != 10 or not cik.isdigit():
-                raise CIKError(cik)
-        elif isinstance(cik, int):
-            if cik > 10**10:
-                raise CIKError(cik)
-            elif cik < 10**9:
-                return str(cik).zfill(10)  # pad with zeros if less than 10 digits given
-        return cik
+        # creating CIK object should check to see if ciks are valid
```

Comment:
Do you think we should do type check for CIK? This section looks a little complicated to me as we are checking the instance type at the runtime level.

One design which I can think of is:
`CIKService`: We can create a separate class name CIKService whose responsibility would be to fetch the particular ticker value from either company name or list of string etc and return the code or codes.

Now, this class will be passed to `Filing`. Basically `Filing` doesn't need to know about how CIK's are fetched, parsed or validated. It should be like a BlackBox to Filing.


For e.g. 

```
class CIKService(): 

    __init__(base_url: String):
        self.base_url = base_url

      def code_for(company_name):
            // fetch the company code
            // parse using the beautifulSoup as you already did
            // validate the code
           return codes
       


Also, if you feel like it is too much change, we can create a separate ticket to handle the design.

Reply:
A lot of times, looking up company name returns multiple search results (ex: lookup "Apple"). Looking up tickers only returns a single CIK when testing on popular tickers (ex: AAPL). Should the default behavior for handling the case where multiple search results are returned be that the first company in the list is given and the user is warned that the CIK may not be correct? Or should it raise an exception and list all of the company names returned when searching for that keyword (like "Apple"), asking the user to choose the exact company name that they want?

Reply:
IMO, if we make it interactive i.e. asking the user to verify the CIK then it will be a bottleneck for the power user. I guess this functionality should not reside on the API level.

Regarding handling edge cases, what we can do is, show the warning to the user that returned CIK code may be wrong and so, please verify on their own or provide the CIK directly and use other API. 

This is not a great user experience but at the same time, we are dependent on some other service to fetch the result whose accuracy we can't guarantee for time-being.